walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,54 @@
 // limitations under the License.
 
 package com.google.sps;
+import com.google.sps.FindMeetingQuery;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
+import java.util.*;
 
-import java.util.Collection;
 
 public final class FindMeetingQuery {
+    Collection<TimeRange> timeRanges = new ArrayList<>();
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return timeRanges;
+    }
+    if(request.getAttendees ().isEmpty() || request.getDuration() == 0 || events.isEmpty()) {
+        timeRanges.add(TimeRange.WHOLE_DAY);
+        return timeRanges;
+    }
+    HashSet<String> req_attendants = new HashSet<>();
+    for(String person: request.getAttendees())
+        req_attendants.add(person);
+    long req_duration = request.getDuration();
+
+    int start = 0;
+    int end = 0;
+    while(end <= TimeRange.END_OF_DAY){
+        if(available(events, req_attendants, end)) {
+            while(available(events, req_attendants, end) && end <= TimeRange.END_OF_DAY)
+                end+=1;
+            if(end-start >= req_duration) {
+                TimeRange time = TimeRange.fromStartEnd(start, end-1, true);
+                timeRanges.add(time);
+            }
+        } else {
+            while(!available(events, req_attendants, end) && end <= TimeRange.END_OF_DAY)
+                end+=1;
+        }   start=end-1;
+    }
+    return timeRanges;
+  }
+  private static boolean available(Collection<Event> events, Collection<String> req_attendants, int timeSlot) {
```

Comment:
I think this helper function is great, it's defined at the right level where it both (a) abstracts away a concrete step in the calculation and (b) is reused multiple times. one suggestion to make it even better would be to add a documentation string before the function declaration, similar to TimeRange.fromStartEnd.

btw while you're at it, I'd say even the public query() method could use some javadoc.